--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -5310,12 +5310,6 @@ another_round:
 		}
 	}
 
-	if (eth_type_vlan(skb->protocol)) {
-		skb = skb_vlan_untag(skb);
-		if (unlikely(!skb))
-			goto out;
-	}
-
 	fast_recv = rcu_dereference(athrs_fast_nat_recv);
 	if (fast_recv) {
 		if (fast_recv(skb)) {
@@ -5324,6 +5318,12 @@ another_round:
 		}
 	}
 
+	if (eth_type_vlan(skb->protocol)) {
+		skb = skb_vlan_untag(skb);
+		if (unlikely(!skb))
+			goto out;
+	}
+
 	if (skb_skip_tc_classify(skb))
 		goto skip_classify;
 
