diff --git a/include/net/route.h b/include/net/route.h
index 2551f3f..e4ca5d0 100644
--- a/include/net/route.h
+++ b/include/net/route.h
@@ -234,6 +234,9 @@ struct rtable *rt_dst_alloc(struct net_device *dev,
 			     bool nopolicy, bool noxfrm);
 struct rtable *rt_dst_clone(struct net_device *dev, struct rtable *rt);
 
+int ip_rt_register_notifier(struct notifier_block *nb);
+int ip_rt_unregister_notifier(struct notifier_block *nb);
+
 struct in_ifaddr;
 void fib_add_ifaddr(struct in_ifaddr *);
 void fib_del_ifaddr(struct in_ifaddr *, struct in_ifaddr *);
